platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/hash/MerkleHashBuilder.java

```diff
@@ -142,77 +113,104 @@ public Future<Hash> digestTreeAsync(MerkleNode root) {
-
-        return () -> {
-            final MerkleIterator<MerkleNode> it = root.treeIterator()
-                    .setFilter(MerkleHashBuilder::filter)
-                    .setDescendantFilter(MerkleHashBuilder::descendantFilter);
-            if (threadId > 0) {
-                // One thread can iterate in-order, all others will use random order.
-                it.setOrder(POST_ORDERED_DEPTH_FIRST_RANDOM);
-            }
-
-            try {
-                hashSubtree(it, activeThreadCount);
-
-                // The last thread to terminate is responsible for setting the future
-                int remainingActiveThreads = activeThreadCount.getAndDecrement() - 1;
-                if (remainingActiveThreads == 0) {
-                    result.set(root.getHash());
-                }
-            } catch (final Throwable t) {
-                result.cancelWithException(t);
-            }
-        };
-    }
+    FutureMerkleHash resultFuture;
 
     /**
      * The root of a merkle tree.
      *
      * @param it
      * 		An iterator that walks through the tree.
-     * @param activeThreadCount
-     * 		If single threaded then this is null, otherwise contains the number of threads
-     * 		that are actively hashing. Once the active thread count dips below the maximum,
-     * 		this means that one thread has either finished or exploded.
      */
-    private void hashSubtree(final Iterator<MerkleNode> it, final AtomicInteger activeThreadCount) {
+    private void hashSubtree(final Iterator<MerkleNode> it) {
         while (it.hasNext()) {
             final MerkleNode node = it.next();
-            // Potential optimization: if this node is currently locked, do not wait here. Skip it and continue.
-            // This would require a lock object that support the "try lock" paradigm.
-            synchronized (node) {
-                if (activeThreadCount != null && activeThreadCount.get() != cpuThreadCount) {
-                    break;
-                }
 
-                if (node.getHash() != null) {
-                    continue;
-                }
+            if (node.getHash() != null) {
+                continue;
+            }
+
+            merkleCryptography.digestSync(node, MERKLE_DIGEST_TYPE);
+        }
+    }
+
+    class TraverseTask extends AbstractTask {
+        final MerkleNode node;
+        final AbstractTask out;
 
-                if (node.isLeaf()) {
-                    merkleCryptography.digestSync(node.asLeaf(), MERKLE_DIGEST_TYPE);
+        TraverseTask(final MerkleNode node, AbstractTask out) {
+            super(threadPool, 1);
+            this.node = node;
+            this.out = out;
+        }
+
+        @Override
+        protected boolean exec() {
+            try {
+                if (node == null || node.getHash() != null) {
+                    out.send();
+                } else if (node.isLeaf()) {
+                    merkleCryptography.digestSync(node, MERKLE_DIGEST_TYPE);
```

Comment:
`ComputeTask` may be used here to make it more consistent with internal nodes, but this is probably an overkill...

Reply:
`ComputeTask` works only with internal nodes and creating another task would just add overhead without any benefit but specifying `node.asLeaf()` totally makes sense here.